plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -0,0 +1,352 @@
+                    total.set(response.getHits().getTotalHits());
+                    onScrollResponse(response);
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    logger.error("failed while executing the initial scroll request", e);
+                    listener.onFailure(e);
+                }
+            });
+        } catch (Throwable t) {
+            finishHim(t);
+        }
+    }
+
+    void onScrollResponse(SearchResponse searchResponse) {
+        threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> {
+            try {
+                scroll.set(searchResponse.getScrollId());
+                SearchHit[] docs = searchResponse.getHits().getHits();
+                logger.debug("scroll returned [{}] documents with a scroll id of [{}]", docs.length, searchResponse.getScrollId());
+                if (docs.length == 0) {
+                    startNormalTermination();
+                    return;
+                }
+                batches.incrementAndGet();
+                List<SearchHit> docsIterable = Arrays.asList(docs);
+                if (mainRequest.size() != -1) {
+                    // Truncate the docs if we have more than the request size
+                    long remaining = max(0, mainRequest.size() - successfullyProcessed());
+                    if (remaining <= docs.length) {
+                        if (remaining < docs.length) {
+                            docsIterable = docsIterable.subList(0, (int) remaining);
+                        }
+                    }
+                }
+                BulkRequest request = buildBulk(docsIterable);
+                if (request.requests().isEmpty()) {
+                    /*
+                     * If we noop-ed the entire batch then just skip to the next
+                     * batch or the BulkRequest would fail validation.
+                     */
+                    startNextScrollRequest();
+                    return;
+                }
+                request.timeout(mainRequest.timeout());
+                request.consistencyLevel(mainRequest.consistency());
+                if (logger.isDebugEnabled()) {
+                    logger.debug("sending [{}] entry, [{}] bulk request", request.requests().size(),
+                            new ByteSizeValue(request.estimatedSizeInBytes()));
+                }
+                client.bulk(request, new ActionListener<BulkResponse>() {
+                    @Override
+                    public void onResponse(BulkResponse response) {
+                        onBulkResponse(response);
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        finishHim(e);
+                    }
+                });
+            } catch (Throwable t) {
+                finishHim(t);
+            }
+        });
+    }
+
+    void onBulkResponse(BulkResponse response) {
```

Comment:
I see. From a design standpoint I'd rather have this as encapsulated as possible (but then I am also one of those guys that tries to declare classes as `final` if possible ;)). 

From a performance standpoint the dispatching via synthetic method might matter (theoretically) but first of all this must be a *very* hot code path and even then the JIT compiler may figure out an optimization so doing this because of performance is highly speculative (esp. without hard numbers gathered in experiments and looking what the JIT made of this).

Personally, I'd lean towards making it private but if having package private access is more common in the code base...